crates/runtime/src/http/v1/nsql.rs

```diff
@@ -42,13 +61,96 @@ fn clean_model_based_sql(input: &str) -> String {
+///  - Distinct columns
+///  - Random sample
+///
+/// Convert the [`SampleTableParams`] into how an LLM would ask to use it (via a [`ChatCompletionRequestAssistantMessage`]).
+/// Convert the result of a [`SampleDataTool`] call how we would return it to the LLM, (via a [`ChatCompletionRequestToolMessage`]).
+async fn sample_messages(
+    sample_from: &[TableReference],
+    df: Arc<DataFusion>,
+) -> Result<Vec<ChatCompletionRequestMessage>, Box<dyn std::error::Error + Send + Sync>> {
+    let mut messages = Vec::with_capacity(4 * sample_from.len());
+
+    for dataset in sample_from {
+        for params in [
+            SampleTableParams::DistinctColumns(DistinctColumnsParams {
+                tbl: dataset.to_string(),
+                limit: 3,
+                cols: None,
+            }),
+            SampleTableParams::RandomSample(RandomSampleParams {
+                tbl: dataset.to_string(),
+                limit: 3,
+            }),
+        ] {
+            let (req, resp) = call_sample_and_create_messages(Arc::clone(&df), &params).await?;
+            messages.push(req.into());
+            messages.push(resp.into());
+        }
+    }
+    Ok(messages)
+}
+
+/// Call the `sample_data` tool with the given parameters and create associated Assistant and Tool messages.
+async fn call_sample_and_create_messages(
+    df: Arc<DataFusion>,
+    params: &SampleTableParams,
+) -> Result<
+    (
+        ChatCompletionRequestAssistantMessage,
+        ChatCompletionRequestToolMessage,
+    ),
+    Box<dyn std::error::Error + Send + Sync>,
+> {
+    let ds = params.dataset();
+    let result = SampleDataTool::new(params.into())
+        .call_with(params, Arc::clone(&df))
+        .await?;
+
+    let req = ChatCompletionRequestAssistantMessageArgs::default()
+        .tool_calls(vec![ChatCompletionMessageToolCall {
+            id: format!("distinct-{ds}-nsql"),
+            r#type: ChatCompletionToolType::Function,
+            function: FunctionCall {
+                name: "sample_data".to_string(),
+                arguments: serde_json::to_string(&params)
+                    .map_err(OpenAIError::JSONDeserialize)?
+                    .to_string(),
+            },
+        }])
+        .build()
+        .boxed()?;
+
+    let resp = ChatCompletionRequestToolMessageArgs::default()
+        .tool_call_id(format!("distinct-{ds}-nsql"))
+        .content(ChatCompletionRequestToolMessageContent::Text(
+            result.to_string(),
+        ))
+        .build()
+        .boxed()?;
+
+    Ok((req, resp))
+}
+
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "lowercase")]
 pub struct Request {
     pub query: String,
 
     #[serde(default = "default_model")]
     pub model: String,
+
+    #[serde(default = "default_sample_data_enabled")]
+    pub sample_data_enabled: bool,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub datasets: Option<Vec<String>>,
```

Comment:
Do we expect users to be able to exclude this property entirely in their requests? If so, this needs a `serde(default, ...)`

Reply:
I don't think so?
```shell
curlie -XPOST -H "Content-Type: application/json" "http://localhost:8090/v1/nsql" -d '{
  "query": "What is the total sales?"
}'
```